test/publisher/functional_test/runner.spec.js

```diff
@@ -0,0 +1,129 @@
+      spyOn(iniLoader, 'loadFrom').andReturn({
+        default: scenarioConfiguration.sharedConfigFile || {}
+      });
+
+      const defaultConstructParams = getConstructParam(defaultConfiguration);
+
+      process.env = Object.assign({}, process.env, scenarioConfiguration.environmentVariables);
+
+      const resolvedCSMConfig = csmConfigProvider();
+      AWS.Service.prototype.publisher = new Publisher(resolvedCSMConfig);
+
+      //start a subprocess to echo back the udp datagrams
+      let fakeAgent = await agentStart(resolvedCSMConfig.port || 31000);
+
+      // listen to the monitoring events echo-ed back
+      let monitoringEvents = [];
+      fakeAgent.stdout.on('data', function(data) {
+        const str = data.toString('utf8');
+        const parsed = str.trim().split('\n').map(JSON.parse);
+        monitoringEvents = monitoringEvents.concat(parsed);
+      })
+
+      //start call according to test cases
+      for (const apiCall of scenario.apiCalls) {
+        const Klass = getServiceClass(apiCall.serviceId);
+        const client = new Klass(defaultConstructParams);
+        if (scenarioConfiguration.accessKey) {
+          client.config.credentials.accessKeyId = scenarioConfiguration.accessKey;
+        }
+        if (scenarioConfiguration.sessionToken) {
+          client.config.credentials.sessionToken = scenarioConfiguration.sessionToken;
+        }
+        const operation = apiCall.operationName[0].toLowerCase() + apiCall.operationName.substring(1);
+        const request = client.makeRequest(operation, apiCall.params);
+        mockResponses(apiCall.attemptResponses, request);
+        request.send();
+      }
+
+      spyOn(iniLoader, 'loadFrom').andReturn(defaultConfiguration.sharedConfigFile || {});
+      //close agent and then validate the datagram echo-ed back from agent
+      await agentFinished(fakeAgent);
+      // console.log('____', monitoringEvents);
+      // console.log('||||', scenario.expectedMonitoringEvents);
+      expect(validateEvents(monitoringEvents, scenario.expectedMonitoringEvents)).to.eql(true);
+    })
+  }
+
+})
+
+function getConstructParam(configurations) {
+  let constructParams = {};
+  if(configurations.userAgent) constructParams.customUserAgent = configurations.userAgent;
+  if(configurations.region) constructParams.region = configurations.region;
+  if(configurations.paramValidation === false) constructParams.paramValidation = configurations.paramValidation;
+  if(configurations.clientSideMonitoring) constructParams.clientSideMonitoring = configurations.clientSideMonitoring;
+  return constructParams;
+}
+
+
+function agentStart(port) {
+   //start up an udp agent echoing back monitoring events
+   //pass in --inspect={port} to debug the subprocess
+  return new Promise((resolve) => {
+    port = port || '';
+    const fakeAgent = fork(path.join(__dirname, 'utils/mock_agent.js'), [`port=${port}`], {stdio: ['pipe', 'pipe', 'pipe', 'ipc']});
```

Comment:
Minor: you'll want to pass `utils` and `mock_agent.js` as separate items for windows compat.